language/move-stdlib/sources/ascii.move

```diff
@@ -3,9 +3,7 @@
 module std::ascii {
     use std::vector;
     use std::option::{Self, Option};
-
-    /// An invalid ASCII character was encountered when creating an ASCII string.
-    const EINVALID_ASCII_CHARACTER: u64 = 0;
```

Comment:
Is it possible to include a custom error message in the error? We can have the err native functions take that error message and print it out along with the stacktrace.

Reply:
I think the developer flow with errmap is terrible, but you may have a different viewpoint. Note that as I say in the PR description, you can still do the errmap thing in your app, just not for those few cases in the stdlib.

I did this PR based on @sblackshear input in this [discord thread](https://discord.com/channels/964205366541963294/964207317430202368/994087514174406696), where he says he is OK with this.

Reply:
> Is it possible to include a custom error message in the error? We can have the err native functions take that error message and print it out along with the stacktrace.

I think it could be sufficient to put a comment into the code:

```
      // The XYZ condition must be true to proceed.
      assert!(cond, code)
```      

Then our (aptos specific) stack trace code can try to grep this from the source.

Notice that the legacy error reason model does just the same, only with more boilerplate:

```
/// Happens if the XYZ condition is not true
const EXYZ_NOT_TRUE: u64 = 3;
...
assert!(cond, EXYZ_NOT_TRUE)
```

There is tooling which extracts the string at the const declaration and puts it into a so-called error map.

Reply:
I think I misunderstood the proposal from our discussion in the contributors chat--I was expecting something more along the lines of:

```
// in ascii.move

// a written spec somewhere says 2001 is the generic category for INVALID_ARGUMENT,
// whereas the const name and error message give the reader a bit more module-specific info
#[err(msg = "An invalid ASCII character was encountered when creating an ASCII string")]
const EInvalidAsciiCharacter = 2001;
```

or

```
// in ascii.move

// a written spec somewhere says 2001 is the generic category for INVALID_ARGUMENT,
// and perhaps there's compiler or linter support for checking that a error-style constant
// with the name EInvalidArgument is always bound to 2001 and not (e.g.) 17.
#[err(msg = "An invalid ASCII character was encountered when creating an ASCII string")]
const EInvalidArgument = 2001;
```

rather than going back to the old way of a dedicated `errors` module w/ function calls to use the error codes in clients.

Reply:
My main desire here is to
- come up with a toolchain-supported discipline for user-defined, application-specific errors
- use this discipline in the stdlib as well so the toolchain only needs to understand one way of doing things

I am by no means wedded to the errmap approach to toolchain support, which I think has a lot of problems (e.g., parsing err messages from comments whereas they should probably live somewhere more structured and tool friendly like an annotation)

I think (and let me know where you disagree) that both of the schemes proposed above are compatible with a discpline for application-specific errors, e.g.

```
#[err(msg = "An application-specific error message")]
const EAppSpecificError = ...; // an application-specific error code, or one of the reserved error codes if this is an app-specific variant of one of the generic errors
```

Whereas I do not immediately see how the `errors.move` approach can be extended to application-specific errors. I think either each application will need to define its own custom errors module (which seems a bit heavy/boilerplate-y to me) or will have to use a different convention, which will make the tooling support more difficult.

Reply:
It appears we had a large misunderstanding in the Discord chat. So lets backtrack to solutions which mitigate the fact that you made a breaking change to the error logic which *forces users of the stdlib* into the error reason/errmap model.

I see a few mitigation strategies:

1. Every application maintains a copy of the stdlib, and uses the error model they prefer.
2. Those functions in the stdlib which do abort use globally unique codes (like 2001 or whatever). 
3. We go to the root of the problem and redefine abort handling in the language/VM.

Regards (3), the core problem is that we are conflating error codes with error reasons. You actually only want error reasons and nothing else (and call them application specific errors). An error reason is best represented by a custom string, like your examples show.  So what you really want is:

```
abort "An invalid ASCII character was encountered when creating an ASCII string"
```

Since that is not possible, you use this (IMO awkward und ugly) hack over the constant with an attribute and/or the error map tools heuristic.

What we want is

```
abort errors::invalid_argument(), "An invalid ASCII character was encountered when creating an ASCII string"
```

Maybe we can extend language and VM to support both? So abort can get a code or a string or both.

PS. BTW the changes you made to errors fundamentally breaks the prover model. If you don't have unique codes, the prover would need to be extended to allow to specify `aborts_if C with Module::ESOME`.

Reply:
I think we can do `abort "An invalid ASCII character was encountered when creating an ASCII string"` if we want without changing the language. I have some ideas, and I can write it up if we like the idea. 

I think I could also pull off `abort errors::invalid_argument(), "An invalid ASCII character was encountered when creating an ASCII string"` if we really want that too :) But its a bit less clean

Reply:
> 2. Those functions in the stdlib which do abort use globally unique codes (like 2001 or whatever).

This seems like the path of least resistance and a good way forward if there is (as I suspect) limited bandwidth on both our sides to do more than this in the short term. 

> 3. We go to the root of the problem and redefine abort handling in the language/VM.

Definitely open to this as well if there's appetite/bandwidth to iterate on the design. It will be more work and take some time, but it would be great if we could get error message strings straight from the bytecode and avoid this dance of needing to find source code in order to report a detailed reason for an error.

Wanted to bring up a few possible points of contention so we can evaluate how much alignment there might be on a first-class approach for string error messages (and feel free to ignore if the preference is to save time by moving forward with (2)):
- I do think it's important to enable using integer codes for space-sensitive users (fine if this just works by stripping the error strings from the bytecode). Hopefully storage costs will be so low that this doesn't matter, but we don't know that this will be true on every platform, and there are often reasons to cut costs even when storage is cheap (and ideally we don't want folks to do that in a hacky way). I ran the #'s and the space impact is nontrival: adding error strings would increase the size of the stdlib by 4% and the DPN by 7%. Not that bad, but enough that we would probably want the option to skip.
- I do think we should try to keep "error output" of the VM as a simple integer code rather than supporting `(code, string)` if we can. Implementation that let you `abort` with strings in source and compile down to `abort i` with some link between `i` and the strong would be one way to preserve this, but there may be others.

Reply:
> PS. BTW the changes you made to errors fundamentally breaks the prover model. If you don't have unique codes, the prover would need to be extended to allow to specify aborts_if C with Module::ESOME.

Yep, this makes sense to me.